drivers/misc/pio_rpi_pico/pio_rpi_pico.c

```diff
@@ -0,0 +1,97 @@
+{
+	const struct pio_rpi_pico_data *data = dev->data;
+
+	return data->state_machine_num;
+}
+
+const uint32_t *const pio_rpi_pico_get_state_machines(const struct device *dev)
+{
+	const struct pio_rpi_pico_data *data = dev->data;
+
+	return data->state_machines;
+}
+
+static int pio_rpi_pico_init(const struct device *dev)
+{
+	const struct pio_rpi_pico_config *config = dev->config;
+	const struct pio_rpi_pico_data *data = dev->data;
+
+	for (size_t i = 0; i < data->state_machine_num; i++) {
+		if (i < config->state_machine_claims_num) {
+			if (pio_sm_is_claimed(config->pio, config->state_machine_claims[i])) {
+				return -EBUSY;
+			}
+			pio_sm_claim(config->pio, config->state_machine_claims[i]);
+			data->state_machines[i] = config->state_machine_claims[i];
+		} else {
+			int sm = pio_claim_unused_sm(config->pio, false);
+
+			if (sm < 0) {
+				return -EBUSY;
+			}
+			data->state_machines[i] = sm;
+		}
+	}
+
+	return pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+}
+
+#define RPI_PICO_PIO_INIT(idx)                                                                     \
+	PINCTRL_DT_INST_DEFINE(idx);                                                               \
+	static const uint8_t sm_claims##idx[] =                                                    \
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(idx, state_machine),                             \
+			    (DT_INST_PROP(idx, state_machine)), ({}));                             \
+	static uint32_t sm##idx[MAX((COND_CODE_1(DT_INST_NODE_HAS_PROP(idx, state_machine_num),    \
+						 (DT_INST_PROP(idx, state_machine_num)), (0))),    \
+				    (COND_CODE_1(DT_INST_NODE_HAS_PROP(idx, state_machine),        \
+						 (ARRAY_SIZE(sm_claims##idx)), (0))))];            \
+	static const struct pio_rpi_pico_config pio_rpi_pico_config_##idx = {                      \
+		.pio = (PIO)DT_INST_REG_ADDR(idx),                                                 \
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),                                       \
+		.state_machine_claims = sm_claims##idx,                                            \
+		.state_machine_claims_num = ARRAY_SIZE(sm_claims##idx),                            \
+	};                                                                                         \
+	static struct pio_rpi_pico_data pio_rpi_pico_data_##idx = {                                \
+		.state_machines = sm##idx,                                                         \
+		.state_machine_num = ARRAY_SIZE(sm##idx),                                          \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(idx, &pio_rpi_pico_init, NULL, &pio_rpi_pico_data_##idx,             \
+			      &pio_rpi_pico_config_##idx, PRE_KERNEL_2,                            \
```

Comment:
Is there a reason for choosing PRE_KERNEL_2 and not earlier? If possible, I think it should be initialized as early as possible, because devices relying on this in the future might need it to be ready very early

Reply:
It makes sense.
I change it to PRE_KERNEL_1, the same as the reset-controller priority.
It is early enough before other driver initialization.